llvm-hs-pure/src/LLVM/IRBuilder/Instruction.hs

```diff
@@ -288,3 +292,32 @@ condBr cond tdest fdest = emitTerm $ CondBr cond tdest fdest []
 -- | See <https://llvm.org/docs/LangRef.html#unreachable-instruction reference>.
 unreachable :: MonadIRBuilder m => m ()
 unreachable = emitTerm $ Unreachable []
+
+-- | Creates a series of instructions to generate a pointer to a string
+-- constant. Useful for making format strings to pass to @printf@, for example
+globalStringPtr
+  :: (MonadModuleBuilder m, MonadIRBuilder m)
+  => String       -- ^ The string to generate
+  -> Name         -- ^ Variable name of the pointer
+  -> m Operand
+globalStringPtr str nm = do
+  let asciiVals = map (fromIntegral . ord) str
+      llvmVals  = map (C.Int 8) (asciiVals ++ [0]) -- append null terminator
+      char      = IntegerType 8
+      charStar  = ptr char
+      charArray = C.Array char llvmVals
+      ty        = LLVM.AST.Typed.typeOf charArray
+  emitDefn $ GlobalDefinition globalVariableDefaults
+    { name                  = nm
+    , LLVM.AST.Global.type' = ty
+    , linkage               = External
+    , isConstant            = True
+    , initializer           = Just charArray
+    , unnamedAddr           = Just GlobalAddr
+    }
+  let address = ConstantOperand $ C.GlobalReference (ptr ty) nm
+      inBounds = True
+      indices = join (++) [ConstantOperand (C.Int 32 0)]
```

Comment:
Please don’t use join here. It might seem like a clever idea right now but it makes the code harder to read and understand. I would just go for `[ConstantOperand (C.Int 32 0), ConstantOperand (C.Int 32 0)]` or if you really want to avoid the duplication bind it in a `let`.

Reply:
If I'm going to add a test case, I'd like to also add a way to verify the validity of the IR generated by the test cases, hopefully in a more astute way then piping llvm:Pretty to lli, any ideas ?

Reply:
Adding a testcase that verifies the validity is slightly tricky as the irbuilder is part of `llvm-hs-pure` which does deliberately not depend on the LLVM C libraries or tooling. So you’ll have to make a test in `llvm-hs`. There are tons of tests that check that we can convert from the Haskell AST to the C++ representation and then print the IR so it should be fairly easy to use one of those as an example.

Reply:
You seem to have deleted your comment asking whether the `llvm-hs-pure` test suite can depend on `llvm-hs` but I’ll respond anyway in case the comment got lost by accident: I’d like to avoid that. Circular dependencies on a package level (even if on a component level there is no cycle) cause problems for some build tools, e.g., nix. Since we control both `llvm-hs` and `llvm-hs-pure` I don’t think there is anything to be gained by adding a test suite to `llvm-hs-pure` that depends on `llvm-hs` and instead I would suggest that you add the test to the test suite of `llvm-hs`.